src/v/raft/state_machine_manager.h

```diff
@@ -238,7 +248,12 @@ class state_machine_manager_builder {
         return {raft, std::move(_stms), _sg};
     }
 
+    const cluster::topic_configuration* initial_topic_cfg() const {
```

Comment:
Raft should never depend on the types from `cluster` submodule. Why generic `raft::state_machine_manager` has to be aware of the topic ? I think it should not be

Reply:
It does not depend on the type as such, as the type is incomplete.

The assumption of a state machine factory being capable to create state machines from consensus instance only is not correct and leads to race conditions we are fixing here. We need a way to pass topic config to the state machine constructor. 

The only alternative I can see would be to add a parameter to `state_machine_factory::create` method and pass it to all STMs whether they need it or not. Would you prefer it?

Also, whether we do the above change or not, I can wrap topic config pointer into a less specifically named class, e.g. `cluster::stm_init_params` to avoid adding more logic into this area should we need to pass something else in future. Do you think it's a good idea?